File: projects/21-bridge-structural-health/fog/aggregation.py
"""Pure window-aggregation math over the (value, ts) pairs produced by
buffering.group_by_key() for one (sensor_type, site_id) group."""

from statistics import fmean


def aggregate(sensor_type, site_id, unit, pairs, window_start, window_end):
    """Reduce one window's (value, ts) pairs into a summary record.

    latest is the last pair in arrival order, not the max-timestamp pair --
    group_by_key() preserves the order readings were appended to RAW in, and
    sensors dispatch batches in chronological order, so arrival order
    already reflects reading order without needing to parse/sort ts.
    """
    values = [value for value, _ts in pairs]
    return {
        "sensor_type": sensor_type,
        "site_id": site_id,
        "unit": unit,
        "window_start": window_start,
        "window_end": window_end,
        "count": len(values),
        "min": min(values),
        "max": max(values),
        "avg": round(fmean(values), 3),
        "latest": values[-1],
    }

File: projects/26-beehive-apiary-monitoring/fog/aggregation.js
"use strict";

function condenseHiveWindow(sensorType, siteId, unit, readings, windowStart, windowEnd) {
  const values = readings.map((r) => r.value);
  const sum = values.reduce((acc, v) => acc + v, 0);
  return {
    sensor_type: sensorType,
    site_id: siteId,
    unit,
    window_start: windowStart,
    window_end: windowEnd,
    count: values.length,
    min: Math.min(...values),
    max: Math.max(...values),
    avg: Math.round((sum / values.length) * 1000) / 1000,
    latest: readings[readings.length - 1].value,
  };
}

module.exports = { condenseHiveWindow };
